Allow disabling caching for webhook authorizers when using `apiserver.config.k8s.io/v1{alpha1,beta1}.AuthorizationConfiguration`

rfranzke opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind feature
/kind api-change

#### What this PR does / why we need it:
~~It is now possible to set the `AuthorizedTTL` or `UnauthorizedTTL` in `apiserver.config.k8s.io/v1{alpha1,beta1}.AuthorizationConfiguration` to `0` in order to disable caching of authorizer decisions. Previously, setting them to `0` caused the default values to take effect.~~

As discussed in https://github.com/kubernetes/kubernetes/pull/129237#issuecomment-2591263513, this PR introduces new boolean `Cache{Una,A}uthorizedRequests` fields in the `apiserver.config.k8s.io/v1{alpha1,beta1}.AuthorizationConfiguration` APIs in order to allow disabling caching.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubernetes/issues/129233

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
kube-apiserver: Caching of authorization webhook decisions for authorized and unauthorized requests can now be disabled in the `--authorization-config` file by setting the new fields `cacheAuthorizedRequests` or `cacheUnauthorizedRequests` to `false` explicitly. See https://kubernetes.io/docs/reference/access-authn-authz/authorization/#using-configuration-file-for-authorization for more details.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

seans3 commented:
/ok-to-test
/sig-auth
/remove-sig api-machinery

seans3 commented:
/sig auth

enj commented:
/hold

This changes the meaning of existing manifests for a v1 / GA feature, as both `0` and `unset` meant "use the default value."

We can probably use any negative value as an indicator to disable the cache.  Another potential approach could be to make the behavior change behind a feature gate so that emulated versions preserve it.

@liggitt probably has thoughts.

liggitt commented:
yeah, the existing type serializes a zero value to an explicit `0`, which we wouldn't want to start interpreting as "do not cache"

I guess I'd ask if a remote authenticator with no cache at all is _really_ something we want / need... not even a `1s` or `1ms` debounce cache? *If* it is (I'm not sure it is), I'd push in the direction of explicit `cacheAuthorizedRequests: false` / `cacheUnauthorizedRequests: false` fields to disable

rfranzke committed:
Introduce new boolean `cache{Una,A}uthorizedRequests` field

rfranzke committed:
Run `hack/update-codegen.sh`

rfranzke committed:
Respect legacy flags values for caching

With the legacy `--authorization-webhook-cache-{un}authorized-ttl`
flags, caching was disabled when the TTL was set to `0`, so let's
continue doing so when building the authz configuration struct.

rfranzke committed:
Pass TTL=0 to webhook authz plugin when cache disabled

fedebongio commented:
/triage accepted
